### PR TITLE
SNPsplit

### PR DESCRIPTION
**Pull request by ewels**
New SNPsplit module. Continuation of #593 by @dpryan79 

'cc @FelixKrueger and @vivekbhr

### CODE REVIEW

**Comment by ewels** [comment]
Hi all, I've just done some initial work on the existing code:

* Fixing linting errors by adding in missing configurations vars, linking up docs into readme etc.
* Renamed most of the bargraph categories to make them a lot shorter
* Reordered bargraph categories to have a more logical order
* Rewrote plot help to make it easier to scan and to list all categories
* Fixed error where MultiQC would generate a report even if no files found, as module wasn't raising a `UserWarning` when it found no data.

Next up I'll work on refactoring the code some more to be a bit more MultiQC-ey. After that I'll look into adding new parsing for the YAML format.

Phil

**Comment by ewels** [comment]
@FelixKrueger - is this still the correct YAML format? https://github.com/ewels/MultiQC_TestData/commit/c6d57080fef9a98718a560ba800ddbd9d970c912

**Comment by FelixKrueger** [comment]
Yes, I think it hasn't changed since then. Many thanks for getting this going!

Cheers, Felix

File: multiqc/modules/snpsplit/snpsplit.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+            regex_patterns = [
+                # Allele-tagging report
+                ['tagging_skipped', r"Reads were unaligned and hence skipped: (\d+)"],
+                ['tagging_unassignable', r"(\d+) reads were unassignable"],
+                ['tagging_genome1', r"(\d+) reads were specific for genome 1"],
+                ['tagging_genome2', r"(\d+) reads were specific for genome 2"],
+                ['tagging_unassignableCT', r"(\d+) reads that were unassignable contained C>T SNPs"],
+                ['tagging_ambiguous', r"(\d+) reads did not contain one of the expected bases"],
+                ['tagging_conflictingSNPs', r"(\d+) contained conflicting allele-specific SNPs"],
```
Comment (ewels) [inline multiqc/modules/snpsplit/snpsplit.py:87]: @felixkrueger -  I'm trying to match these old categories with the ones in the new YAML. So far I have:

* `tagging_skipped` - `tagging_unaligned`
* `tagging_unassignable` - `tagging_unassignable`
* `tagging_genome1` - `tagging_g1`
* `tagging_genome2` - `tagging_g2`
* `tagging_unassignableCT` - `tagging_CT_positions_skipped`
* `tagging_ambiguous` - **?**
* `tagging_conflictingSNPs` - **??**

Could you help with the last two please?

There are also loads of new categories:

* `tagging_N_containing_reads`
* `tagging_N_deletion`
* `tagging_N_not_known`
* `tagging_N_was_known_SNP`
* `tagging_SNP_annotation`
* `tagging_SNPs_stored`
* `tagging_bizarre`
* `tagging_multi_N_deletion`
* `tagging_no_snp`
* `tagging_non_N_containing_reads`
* `tagging_total_reads`
* `tagging_unassigned_but_ct`

Should I plot all of these as well? Do you have any documentation for these? The docs don't seem to mention them and there hasn't been a release for a long time 😉 

If you could write one-line descriptions for each category that would be great, and I'll add them to the help text for the plot 👍

Comment (ewels) [reply multiqc/modules/snpsplit/snpsplit.py:87]: The reason I ask is that if I can save the fields from the old report style using the same keys as in the new reports, then all of the plots etc should work the same for both outputs.

File: multiqc/modules/snpsplit/snpsplit.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                ['sorting_unassignableReads', "Reads were unassignable"],
+                ['sorting_unassignableReads', "Read pairs were unassignable (UA/UA):"],
+                ['sorting_genome1Reads', "Reads were specific for genome 1"],
+                ['sorting_genome1Reads', "Read pairs were specific for genome 1 (G1/G1)"],
+                ['sorting_genome2Reads', "Reads were specific for genome 2"],
+                ['sorting_genome2Reads', "Read pairs were specific for genome 2 (G2/G2)"],
+                ['sorting_G1UA', "Read pairs were a mix of G1 and UA"],
+                ['sorting_G2UA', "Read pairs were a mix of G2 and UA"],
+                ['sorting_G1G2', "Read pairs were a mix of G1 and G2"],
+                ['sorting_conflictingReads', "Read pairs contained conflicting SNP information"]
```
Comment (ewels) [inline multiqc/modules/snpsplit/snpsplit.py:108]: Need to do the same matching up here...

* `sorting_unassignableReads`
* `sorting_genome1Reads`
* `sorting_genome2Reads`
* `sorting_G1UA`
* `sorting_G2UA`
* `sorting_G1G2`
* `sorting_conflictingReads`

New fields:

* `sorting_PE_conflicting`
* `sorting_PE_conflicting_pairs`
* `sorting_PE_conflicting_singletons`
* `sorting_PE_genome1`
* `sorting_PE_genome1_pairs`
* `sorting_PE_genome1_singletons`
* `sorting_PE_genome2`
* `sorting_PE_genome2_pairs`
* `sorting_PE_genome2_singletons`
* `sorting_PE_total_pairs`
* `sorting_PE_total_reads`
* `sorting_PE_total_singletons`
* `sorting_PE_unassignable`
* `sorting_PE_unassignable_pairs`
* `sorting_PE_unassignable_singletons`

If this gets too difficult / complicated I guess we could also just ditch support for the old-style logs.. How long were they around for? How important is it to maintain support for them?

Comment (FelixKrueger) [reply multiqc/modules/snpsplit/snpsplit.py:108]: I suppose it is fair enough to say that MultiQC integrations works from the YAML file onwards. Wouldn't you agree?

Comment (ewels) [reply multiqc/modules/snpsplit/snpsplit.py:108]: I'd feel more comfortable saying that if SNPsplit had any stable releases that generated YAML files 😅 (or any documentation about the different fields)

Comment (FelixKrueger) [reply multiqc/modules/snpsplit/snpsplit.py:108]: I would feel more comfortable to make a stable release if I knew that MultiQC would be fine with format... :) How about I will make one as soon as I see the first beautiful plots? Need to attend a campfire in the garden right now, talk later.

Comment (ewels) [reply multiqc/modules/snpsplit/snpsplit.py:108]: Hah, fair enough.. I'll see what I can do 👍

Comment (ewels) [reply multiqc/modules/snpsplit/snpsplit.py:108]: Rules for understanding: trim `PE_` / `SE_` prefix, ignore anything that is `_pairs` or `_singletons` (key without suffix is the total). Then we have fewer fields to match and it gets easier:

* `sorting_PE_conflicting`
* `sorting_PE_genome1`
* `sorting_PE_genome2`
* `sorting_PE_total_pairs`
* `sorting_PE_unassignable`

File: multiqc/modules/snpsplit/snpsplit.py (hunk 1 above)
Comment (ewels) [reply multiqc/modules/snpsplit/snpsplit.py]: Ignoring most of these new categories for now. Managed to match up most things making use of a handy crib-sheet from @felixkrueger.

**Comment by ewels** [comment]
Ok, module finished I think. Example report: [multiqc_report.html.zip](https://github.com/ewels/MultiQC/files/4437104/multiqc_report.html.zip)

Module supports both the old verbose log format and also the new YAML format.

Now just waiting for @FelixKrueger to provide some test data run through both the old log format and the new YAML format, to confirm that all of the numbers line up exactly.

**Comment by FelixKrueger** [comment]
> Ok, module finished I think. Example report: [multiqc_report.html.zip](https://github.com/ewels/MultiQC/files/4437104/multiqc_report.html.zip)
> 
> Module supports both the old verbose log format and also the new YAML format.
> 
> Now just waiting for @FelixKrueger to provide some test data run through both the old log format and the new YAML format, to confirm that all of the numbers line up exactly.

just sent them via Slack, can also add them here. Fingers crossed!
[v3.3_reports.zip](https://github.com/ewels/MultiQC/files/4437125/v3.3_reports.zip)

**Comment by ewels** [comment]
Many thanks @FelixKrueger / @dpryan79 - now merged and ready for use! 🚀

**Comment by FelixKrueger** [comment]
Great, thanks so much both of you! Now MultiQC will finally be a really useful tool! 😺

**Comment by dpryan79** [comment]
Thanks everyone!

> Now MultiQC will finally be a really useful tool

LOL

**Pull request merged**